File: Arrays/03-remove_duplicates_sorted_array/C++/better.cpp
#include <iostream>
#include <vector>
using namespace std;

/*
Better Approach
---------------------
INTUITION - create a new vector and iterate through the original vector, add into the result vector if the number is not equl to the last element of the vector
TIME COMPLEXITY - O(N)
SPACE COMPLEXITY - O(N)

*/

vector<int> remove_duplicates_better(vector<int> array) {
    vector<int> res;
    for (int num : array) {
        if (res.empty() || num != res.back()) {
            res.push_back(num);
        }
    }
    return res;
}

int main() {
    vector<int> vec = {1, 1, 2, 2, 2, 3, 3, 3};
    vector<int> res = remove_duplicates_better(vec);
    for (auto n : res){
        cout << n << " ";
    }
    cout << endl;
    return 0;
}

File: Arrays/03-remove_duplicates_sorted_array/Python/better.py
from typing import List

# Better Approach
# ---------------------
# INTUITION - create a new res array, and iterate over the original array, if the number not equal to the last number in the res array then push it in res array else pass.
# TIME COMPLEXITY - O(N)
# SPACE COMPLEXITY - O(N) - result vector

def remove_duplicates_better( array : List[int] ) -> List[int] :
    if not array or len(array) == 1: return array
    res = []
    for number in array:
        if not res or res[-1] != number:
            res.append(number)

    return res
        
# check
arr = [1, 1, 2, 2, 2, 3, 4, 4, 4, 4 ,4, 9, 9]
print(remove_duplicates_better(arr))